src/main/scala/firrtl/transforms/FixAddingNegativeLiteralsTransform.scala

```diff
@@ -0,0 +1,106 @@
+    */
+  def maxNegNumber(width: BigInt): BigInt = -(1 << (width.toInt - 1))
+
+  /** Updates the type of the DoPrim from its arguments (e.g. if is UnknownType)
+    * @param d input DoPrim
+    * @return updated DoPrim with calculated type
+    */
+  def setType(d: DoPrim): DoPrim = {
+    PrimOps.set_primop_type(d)
+  }
+
+  /** Returns a module with fixed additions of negative literals
+    * @param m input module
+    * @return updated module
+    */
+  def fixupModule(m: DefModule): DefModule = {
+    val namespace = Namespace(m)
+    m map fixupStatement(namespace)
+  }
+
+  /** Returns a statement with fixed additions of negative literals
+    * @param namespace object to enabling creating unique names
+    * @param s input statement
+    * @return updated statement
+    */
+  def fixupStatement(namespace: Namespace)(s: Statement): Statement = {
+    val stmtBuffer = mutable.ArrayBuffer[Statement]()
+    val ret = s map fixupStatement(namespace) map fixupExpression(Utils.get_info(s), namespace, stmtBuffer)
+    if(stmtBuffer.isEmpty) {
+      ret
+    } else {
+      stmtBuffer += ret
+      Block(stmtBuffer.toList)
+    }
+  }
+
+  /** Returns a statement with fixed additions of negative literals
+    * @param info Info of statement containing this expression
+    * @param namespace object to enabling creating unique names
+    * @param stmtBuffer mutable buffer of statements - append to this for it to be inlined in the module
+    * @param e expression to fixup
+    * @return fixed expression
+    */
+  def fixupExpression(info: Info, namespace: Namespace, stmtBuffer: mutable.ArrayBuffer[Statement])
+                     (e: Expression): Expression = {
+
+    // Helper function to create the subtraction expression
+    def fixupAdd(expr: Expression, litValue: BigInt, litWidth: BigInt): DoPrim = {
+      if(litValue == maxNegNumber(litWidth)) {
+        val posLiteral = SIntLiteral(-litValue)
+        assert(posLiteral.width.asInstanceOf[IntWidth].width - 1 == litWidth)
+        val sub = DefNode(info, namespace.newTemp, setType(DoPrim(Sub, Seq(expr, posLiteral), Nil, UnknownType)))
+        val tail = DefNode(info, namespace.newTemp, setType(DoPrim(Tail, Seq(WRef(sub)), Seq(1), UnknownType)))
+        stmtBuffer += sub
+        stmtBuffer += tail
+        setType(DoPrim(AsSInt, Seq(WRef(tail)), Nil, UnknownType))
+      } else {
+        val posLiteral = SIntLiteral(-litValue)
+        setType(DoPrim(Sub, Seq(expr, SIntLiteral(-litValue, IntWidth(litWidth))), Nil, UnknownType))
+      }
+    }
+
+    e map fixupExpression(info, namespace, stmtBuffer) match {
+      case DoPrim(Add, Seq(arg, lit@SIntLiteral(value, w@IntWidth(width))), Nil, t: SIntType) if value < 0 =>
+        fixupAdd(arg, value, width)
+      case DoPrim(Add, Seq(lit@SIntLiteral(value, w@IntWidth(width)), arg), Nil, t: SIntType) if value < 0 =>
+        fixupAdd(arg, value, width)
+      case other => other
+    }
+  }
+}
+
+/** Replaces non-expanding arithmetic
```

Comment:
I think this is copypasta from `ReplaceTruncatingArithmetic`?

Reply:
dang I had my suggestions staged but forgot to send them so you beat me to the punch here :)